tests/hypothesis_strategies/utils.py

```diff
@@ -19,18 +29,40 @@ def interval(
         (exclude_bounded, exclude_half_bounded, exclude_fully_unbounded)
     ), "At least one Interval type must be allowed."
 
-    # Create interval from ordered pair of floats
-    bounds = (
-        st.tuples(st.floats(), st.floats()).map(sorted).filter(lambda x: x[0] < x[1])
-    )
-    interval = Interval.create(draw(bounds))
-
-    # Filter excluded intervals
-    if exclude_bounded:
-        assume(not interval.is_bounded)
-    if exclude_half_bounded:
-        assume(not interval.is_half_bounded)
-    if exclude_fully_unbounded:
-        assume(not interval.is_fully_unbounded)
-
-    return interval
+    # Draw the interval type from the allowed types
+    type_gate = {
+        IntervalType.FULLY_UNBOUNDED: not exclude_fully_unbounded,
+        IntervalType.HALF_BOUNDED: not exclude_half_bounded,
+        IntervalType.BOUNDED: not exclude_bounded,
+    }
+    allowed_types = [t for t, b in type_gate.items() if b]
+    interval_type = draw(st.sampled_from(allowed_types))
+
+    # A strategy producing finite floats
+    ffloats = st.floats(allow_infinity=False, allow_nan=False)
+
+    # Draw the bounds depending on the interval type
+    if interval_type is IntervalType.FULLY_UNBOUNDED:
+        bounds = (None, None)
+    elif interval_type is IntervalType.HALF_BOUNDED:
+        bounds = draw(
+            st.sampled_from(
+                [
+                    (None, draw(ffloats)),
+                    (draw(ffloats), None),
+                ]
+            )
+        )
+    elif interval_type is IntervalType.BOUNDED:
+        bounds = draw(
```

Comment:
Looking at the other parts of the code, I would expect that you could also implement this via using `draw(ffloats)` similar to how it is done for the half-bounded intervals. In case this is true, what is the reason to not do it and rely on `hnp` instead?

Reply:
Yes, was my first attempt, in fact. But it turned out the `hnp` version was much more elegant due to the built-in `unique=True` feature, which is crucially needed here. The alternative would have been quite messy, because you'd need to first create the lower bound and then based on it the upper, but the problem is that while creating the lower you still need to ensure that there is still "space" for the upper. So it would have involved some nasty handling of `np.nextafter` etc. The current solution does it in one atomic step, no rejection sampling involved 😎